### PR TITLE
[API] ProductVariant serialization with new key "price"

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

ProductVariant returns additional key "price" which is calculated with ProductVariantPriceCalculator

image

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -55,6 +55,17 @@
             <argument>%sylius.security.new_api_route%</argument>
         </service>
 
+        <service
+                id="sylius.api.swagger_product_variant_documentation_normalizer"
+                class="Sylius\Bundle\ApiBundle\Swagger\ProductVariantDocumentationNormalizer"
+                decorates="api_platform.swagger.normalizer.documentation"
+                public="true"
+                autoconfigure="false"
+                decoration-priority="20"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml:64]:
```suggestion
            id="sylius.api.swagger_product_variant_documentation_normalizer"
            class="Sylius\Bundle\ApiBundle\Swagger\ProductVariantDocumentationNormalizer"
            decorates="api_platform.swagger.normalizer.documentation"
            public="true"
            autoconfigure="false"
            decoration-priority="20"
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Webmozart\Assert\Assert;
+
+final class ProductVariantSerializer implements NormalizerInterface
+{
+    /** @var NormalizerInterface */
+    private $objectNormalizer;
+
+    /** @var ProductVariantPriceCalculator */
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php:18]: Use interface here

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 3]
```diff
@@ -182,19 +180,11 @@ private function hasProductWithPriceInChannel(array $products, int $price, Chann
             foreach ($product['variants'] as $variantIri) {
                 $this->client->executeCustomRequest(Request::custom($variantIri, HttpRequest::METHOD_GET));
 
-                /** @var array $channelPricings */
-                $channelPricings = $this->responseChecker->getValue($this->client->getLastResponse(), "channelPricings");
+                /** @var int $variantPrice */
+                $variantPrice = $this->responseChecker->getValue($this->client->getLastResponse(), "price");
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:184]:
```suggestion
                $receivedPrice = $this->responseChecker->getValue($this->client->getLastResponse(), "price");
```

Comment (SirDomin) [reply src/Sylius/Behat/Context/Api/Shop/ProductContext.php:184]: the value we get is price of a variant i dont see reasons to name it received price :/ maybe receivedVariantPrice?

File: src/Sylius/Bundle/ApiBundle/Swagger/ProductVariantDocumentationNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        $this->decoratedNormalizer = $decoratedNormalizer;
+    }
+
+    public function supportsNormalization($data, string $format = null)
+    {
+        return $this->decoratedNormalizer->supportsNormalization($data, $format);
+    }
+
+    public function normalize($object, string $format = null, array $context = [])
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Swagger/ProductVariantDocumentationNormalizer.php:25]:
```suggestion
    public function normalize($object, $format = null, array $context = [])
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    public function normalize($object, $format = null, array $context = [])
+    {
+        Assert::isInstanceOf($object, ProductVariantInterface::class);
+
+        $data = $this->objectNormalizer->normalize($object, $format, $context);
+        if (isset($data['product'])) {
+            $data['product'] = $data['product']['@id'];
+            $data['translations'] = $this->serializeTranslation($data['translations']);
+        }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php:43]: Why do you need to do that in PR for price serialization?

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        return $data;
+    }
+
+    public function supportsNormalization($data, $format = null, $context = []): bool
+    {
+        return
+            $data instanceof ProductVariantInterface &&
+            $context['operation_type'] !== 'subresource' &&
+            $context['groups'] != 'shop:cart:read' &&
+            $context['groups'] != 'admin:order_item_unit:read';
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php:55]: Maybe you should allow for all shop routes?

Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php]: im still wokring on it, gona refactor before finishing, my goal is to get all tests green

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, @SirDomin! :1st_place_medal:

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Core\Calculator\ProductVariantPriceCalculatorInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Webmozart\Assert\Assert;
+
+final class ProductVariantSerializer implements ContextAwareNormalizerInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php:15]: We are missing spec for this class

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function supportsNormalization($data, $format = null, $context = []): bool
+    {
+        if ($data instanceof ProductVariantInterface) {
+            if (isset($context['item_operation_name']) && $context['item_operation_name'] === 'admin_get') {
+                return false;
+            }
+            return true;
+        }
+        return false;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php:58]: This paragraph can be reduced to this return 
```suggestion
        return 
            $data instanceof ProductVariantInterface &&
            (!isset($context['item_operation_name']) || !$context['item_operation_name'] === 'admin_get')
        ;
```

The next iteration could be extracting to private methods like `isProductVariantInterface` and `isAdminOperation`

nonetheless, we have about 4 cases here to test with spec

**Pull request merged**